### PR TITLE
Add csv format output to llvm-cas-object-format

### PR DESCRIPTION
**Pull request by rastogishubham**
Adrian and I want to be able to output the object stats into a csv format to be able to quickly generate graphs to measure improvements of the changes we make to llvm for the debug-info. I have added the command line option --object-stats-csv to output the object stats in a csv format.

### CODE REVIEW

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (cachemeifyoucan) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I would prefer this write to a file?

stdout is kind of crowded with root id and maybe other things.

**Review by cachemeifyoucan — COMMENTED** [review]
Also please add a test case

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Maybe we can change `object-stats` to take a file argument; passing `-` will have it use `outs()` instead of opening a file (as usual); and then make this option `object-stats-format` that takes either `pretty` or `csv`.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: LLVM style has a space between `if` and `(`. Please run git-clang-format on the patch to catch other  things as well! (Although, see below, I don't think you need this `if`.)

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: You've duplicated everything, which means any time we change the stats we're going to have to remember to change two places.

It seems better to avoid nesting, and instead just customize the format. I think something like this would work:
```
  StringLiteral HeaderFormatPretty = "{0,-22} {1,+10} {2,+7} {3,+10} {4,+7} {5,+10} "
                                "{6,+7} {7,+10} {8,+7} {9,+10} {10,+7}\n";
  StringLiteral FormatPretty = "{0,-22} {1,+10} {2,+7:P} {3,+10} {4,+7:P} {5,+10} "
                          "{6,+7:P} {7,+10} {8,+7:P} {9,+10} {10,+7:P}\n";
  StringLiteral FormatCSV = "{0}, {1}, {3}, {5}, {7}, {9}\n";

  StringRef HeaderFormat = IsCSV ? FormatCSV : HeaderFormatPretty;
  StringRef Format = IsCSV ? FormatCSV : FormatPretty;
  outs() << llvm::formatv(HeaderFormat.begin(), "Kind", "Count", "", "Parents",
                           "", "Children", "", "Data (B)", "", "Cost (B)", "");
   if (!IsCSV)
     outs() << llvm::formatv(HeaderFormat.begin(), "====", "=====", "",
                             "=======", "", "========", "", "========", "",
                             "========", "");
```

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Are the other stats below going to be a problem for the CSV?

**Review by dexonsmith — CHANGES_REQUESTED** [review]
Great idea! I agree with @cachemeifyoucan 's comment, and also have some other feedback.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (cachemeifyoucan) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Code is still quite duplicated here. I wonder if you can just print a `llvm::formatv` CSV output only since that is quite readable too. Just print that to stdout by default or `-` and write to a file when a path is given.

File: llvm/test/tools/llvm-cas-object-format/csv-stats.test
Comment (cachemeifyoucan) [inline llvm/test/tools/llvm-cas-object-format/csv-stats.test:1]: Test looks good. You might be able to just write everything in one file unless you are thinking about reusing `.ll` file something else.

Comment (rastogishubham) [reply llvm/test/tools/llvm-cas-object-format/csv-stats.test:1]: I think it gives me more flexibility if the .ll file was kept separate for now

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I saw duncan's comments after I pushed this change, the code is completely different now

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Thanks duncan, helped me trim down the duplication a lot

**Comment by rastogishubham** [comment]
@dexonsmith @cachemeifyoucan please see the patch now, I have updated the code to minimize duplication and also made it so that it outputs to a csv file. Please let me know if you think there is a cleaner way of doing this

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (cachemeifyoucan) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I will make a proper `llvm::Error` here and `ExitOnError` with proper error message. This should not be an assert since this is a user input error.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (cachemeifyoucan) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: It is a bit weird you want this to be a function instead of just `raw_ostream &`.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (cachemeifyoucan) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: This part can always write to `outs()`

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (cachemeifyoucan) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: This might be fine but technically `StringLiteral` doesn't hold the null terminator but I guess there is always one.

**Review by cachemeifyoucan — COMMENTED** [review]
Looks better. Some comments inline.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I don't know; makes sense to me to have it next to the table.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I'd still like to see the change I suggested earlier:

> Maybe we can change object-stats to take a file argument; passing `-` will have it use `outs()` instead of opening a file (as usual); and then make this option `object-stats-format` that takes either `pretty` or `csv`.

Then we have the following command-lines:
```
# Print stats to stdout
llvm-cas-object-file ... -object-stats -
llvm-cas-object-file ... -object-stats - -object-stats-format=pretty

# Print CSV stats to stdout
llvm-cas-object-file ... -object-stats - -object-stats-format=csv

# Print to a file
llvm-cas-object-file ... -object-stats stats
llvm-cas-object-file ... -object-stats stats -object-stats-format=pretty
llvm-cas-object-file ... -object-stats stats -object-stats-format=csv
```

Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I am working on it

**Comment by rastogishubham** [comment]
@cachemeifyoucan @dexonsmith how does it look now?

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Why is this out-of-line and creating a global variable, instead of being created where it's used?
- If the goal is to package logic into a function, I suggest using a lambda function.
- If the goal is to avoid constructing a `raw_fd_ostream`, I suggest using `Optional<raw_fd_ostream>` and then emplacing it when necessary.
- Or if it's something else, I'm curious!

Comment (dexonsmith) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: (But also, see my other comment below about `raw_fd_ostream`...)

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: LLVM style starts local variables with capital letters. This is also quite common, so the usual name is shortened to the initialize `OS` instead of `OutStream`.

Comment (dexonsmith) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I'm also wondering if this output stream configuration logic could/should be done in `main`, and something like a `StatsOS` passed in as a parameter, rather than doing it here. Have a look and see if you think that's better. That would also allow these options to be declared as local variables in `main` which could be a nice follow-up commit.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: FYI, LLVM style would call this `printToStdout`.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I think it's easier to read with `ComputeStats == "-"`. You might need to wrap in a `StringRef` with `StringRef(ComputStats) == "-"`.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Also, this variable is only used once. I suggest not creating it. Just do the comparison inline.

**Review by dexonsmith — CHANGES_REQUESTED** [review]
Looking pretty good! Just a bunch of style nitpicks now.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I will be honest, I did this because that is how `outs() ` is implemented, and I didn't realize that the `raw_fd_ostream` it creates is static, thanks for pointing it out.

I didn't know of `Optional` so I can check it out.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Good idea, let me see what that looks like

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Did you mean `PrintToStdout`?

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Thanks, I can do that

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Probably the parameter can be a `raw_ostream` since none of the users care that it's related to file descriptor.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Hah, yeah, funny typo :).

**Comment by cachemeifyoucan** [comment]
Also, target branch `experimental/cas/main` is setup with swift-ci test. Feel free to use that.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I didn't make this change because main calls to `computeStats`, which then calls `printToOuts` So I would have to thread the argument through two function calls, with it having no use in `computeStats`

But that is just my preference, please let me know if you disagree, and I will make the change

Comment (dexonsmith) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: That doesn't seem too bad. Alternatively, you could refactor to make `computeStats` to return a result, and then call a second function to print them.

Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I went with the former!

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Now that this is not short, the condition should inverted to reduce nesting:
```
if (ComputeStats.empty())
  return 0;

bool PrintToStdout = false;
// etc.
return 0;
```

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I don't think you use the `fd` part of this. OS should just be `raw_ostream`.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: I think there's a `using namespace llvm` at the top of this file, so you can just say `Optional` here.

**Comment by cachemeifyoucan** [comment]
LGTM now. Please rebase on top of the branch to resolve conflict

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: This is used not just for `.csv` files so it should be called something else, such as `StatsFile`.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Checking `PrintToStdout` again seems unnecessary / a bit brittle.

I suggest:
```
raw_ostream *StatsOS = &outs();
Optional<raw_fd_ostream> StatsFile;
if (!PrintToStdout) {
  std::error_code EC;
  StatsFile.emplace(...);
  ExitOnErr(errorCodeToError(EC));
  StatsOS = &*StatsFile;
}
computeStats(..., *StatsOS);
```

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [inline llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: A bit confusing to create a new `ExitOnError` here, since there's already one with the same name (`ExitOnErr`) in the same scope. And also, this one seems to lack a banner.

If you don't have a specific reason to create a new one, I suggest using the one that already exists.

**Review by dexonsmith — CHANGES_REQUESTED** [review]
It looks like you still need to rebase against ToT, since GitHub is reporting conflicts on the branch.

Also, please update the commit message. Looking at https://github.com/apple/llvm-project/commit/c6dc0960fcb10fcb8bc90bc0a7e2de328d0fcd4b, GitHub is summarizing it as:
```
Add option --object-stats-csv, that output the object statistics in a…
… csv format to a csv file
```
so I suspect the raw text is:
```
Add option --object-stats-csv, that output the object statistics in a csv format to a csv file
```
- This isn't accurate (anymore!). Please update!
- It doesn't say what you're adding the option to. Is this being added to `clang`, `opt`, ...?
- The first paragraph is a bit too long; ideally, it should be short/concise enough that GitHub doesn't feel the need to cut it off; more importantly, we want `git log --oneline` to give good information even with fairly small terminal settings.small terminal settings. This can just be a summary.
- Usually you want some other paragraphs (separated by blank lines) for a good commit message to explain what might not be obvious from a one-liner.
- Reflow all paragraphs (except the first) to be no bigger than 80-columns (many people use 72-columns) so that `git log` is consistent  / easy to read even when the terminal has lots of columns.

I suggest something like:
```
CAS.o: Add -object-stats-format=csv and an argument to -object-stats

Add `-object-stats-format=csv` to llvm-cas-object-format. This changes the
output format of `-object-stats` to comma-separated.

As a drive-by, change `-object-stats` to require an argument for where to
send the stats. This allows the stats to be saved separately from the other
output.
```
Feel free to embellish and/or reword! But this contains the minimum information I think.

BTW, it would have been cleaner to split this into two incremental patches: one for changing `-object-stats` and another for adding the new option. If this were upstream I might have asked for that, but at this point (on this branch) I think it's fine to leave these squashed into this one PR.

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (dexonsmith) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Marked as unresolved, since it looks like the most recent commit didn't make a change here. Maybe you pushed the wrong one?

**Comment by rastogishubham** [comment]
@dexonsmith I wanted to resolve the merge conflict after all the minor changes and updates have been made. I will update the commit message before I merge the PR as well. I just wanted to get all the code changes in first. 

Thanks for the feedback

File: llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp
Comment (rastogishubham) [reply llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp]: Sorry about that, it is fixed now

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/test/tools/llvm-cas-object-format/csv-stats.test
Comment (adrian-prantl) [inline llvm/test/tools/llvm-cas-object-format/csv-stats.test]: It might be easier to debug the test if you use different output files for each RUN line

**Review by dexonsmith — APPROVED** [review]
> @dexonsmith I wanted to resolve the merge conflict after all the minor changes and updates have been made. I will update the commit message before I merge the PR as well. I just wanted to get all the code changes in first.

New commit message LGTM! Feel free to merge once tests pass.

**Comment by rastogishubham** [comment]
@swift-ci please test

**Pull request merged**
